hivemind/client/optim/collaborative.py

```diff
@@ -0,0 +1,283 @@
+    :param target_group_size: maximum group size for DecentralizedAverager's all-reduce
+    :param min_refresh_period: wait for at least this many seconds before fetching new collaboration state
+    :param max_refresh_period: wait for at most this many seconds before fetching new collaboration state
+    :param default_refresh_period: if no peers are detected, attempt to fetch collaboration state this often (seconds)
+    :param expected_drift_peers: assume that this many new peers can join between steps
+    :param expected_drift_rate: assumes that this fraction of current collaboration can join/leave between steps
+    :note: the expected collaboration drift parameters are used to adjust the frequency with which this optimizer will
+      refresh the collaboration-wide statistics (to avoid missing the moment when to run the next step)
+    :param bandwidth: peer's network bandwidth for the purpose of load balancing (recommended: internet speed in mbps)
+    :param performance_ema_alpha: smoothing value used to estimate this peer's performance (training samples per second)
+    :param averaging_expiration: peer's requests for averaging will be valid for this many seconds
+    :param metadata_expiration: peer's metadata (e.g. samples processed) is stored onto DHT for this many seconds
+    :param averaging_timeout: if an averaging step hangs for this long, it will be cancelled.
+    :param scheduler: if specified, use this scheduler to update optimizer learning rate
+    :note: if you are using CollaborativeOptimizer with a lr_scheduler, it is recommended to pass this scheduler
+      explicitly into this class. Otherwise, scheduler may not be synchronized between peers.
+    """
+
+    def __init__(
+            self, opt: torch.optim.Optimizer, *, dht: DHT, prefix: str, target_group_size: int, target_batch_size: int,
+            batch_size_per_step: Optional[int] = None, scheduler: Optional[LRSchedulerBase] = None,
+            min_refresh_period: float = 0.5, max_refresh_period: float = 30, default_refresh_period: float = 3,
+            expected_drift_peers: float = 3, expected_drift_rate: float = 0.2, performance_ema_alpha: float = 0.1,
+            metadata_expiration: float = 30.0, averaging_timeout: Optional[float] = None, verbose: bool = False,
+            **kwargs):
+        super().__init__(opt, dht)
+        self.prefix, self.scheduler = prefix, scheduler
+        self.target_batch_size, self.batch_size_per_step = target_batch_size, batch_size_per_step
+        self.min_refresh_period, self.max_refresh_period, self.default_refresh_period =\
+            min_refresh_period, max_refresh_period, default_refresh_period
+        self.expected_drift_peers, self.expected_drift_rate = expected_drift_peers, expected_drift_rate
+        self.averaging_timeout, self.metadata_expiration = averaging_timeout, metadata_expiration
+        self.averager = TrainingAverager(
```

Comment:
Maybe we should accept averager class in constructor or give a possibility to override averager in subclass?
I don't see an easy way to change averaging logic (e.g. average opt stats) without copypasting the whole class. Correct me if I'm wrong

Reply:
Good point, that might prove useful in subsequent research for secure/robust averagers
/* Added averager_cls parameter that allows user to override the averager class */

Reply:
Conversation log with @nevec : we have instead opted for a _make_averager method for more versatility